internal/base/base.go

```diff
@@ -655,9 +655,101 @@ func DecodeSchedulerEnqueueEvent(b []byte) (*SchedulerEnqueueEvent, error) {
 	}, nil
 }
 
+// Deadlines is a collection that holds deadlines for all active tasks.
+// When a deadline is reached, the associated cancel function is executed.
+// Assumes that deadlines have unique IDs and will not be added multiples times.
+//
+// Deadlines is safe for concurrent use by multiple goroutines.
+type Deadlines struct {
+	update chan deadline
+}
+
+// NewDeadlines returns a Deadlines instance.
+func NewDeadlines(abort chan struct{}, size int) *Deadlines {
+	d := &Deadlines{update: make(chan deadline, size)}
+	go func() {
+		dls := []deadline{}
+		for {
+			if func() bool {
+				var c <-chan time.Time
+				if len(dls) == 0 {
+					c = make(chan time.Time) // dummy
+				} else {
+					t := time.NewTimer(time.Until(dls[0].dl))
+					defer t.Stop()
+					c = t.C
+				}
+				select {
+				case <-abort:
+					return true
+				case dl := <-d.update:
+					if dl.fn != nil {
+						dls = insertDeadline(dls, dl)
+					} else {
+						dls = removeDeadline(dls, dl)
+					}
+				case <-c:
+					dl := dls[0]
+					dls = removeDeadline(dls, dl)
+					dl.fn()
+				}
+				return false
+			}() {
+				return
+			}
+		}
+	}()
+	return d
+}
+
+// Add adds a new cancel func to the collection.
+func (d *Deadlines) Add(id string, dl time.Time, fn context.CancelFunc) {
+	d.update <- deadline{id: id, dl: dl, fn: fn}
+}
+
+// Delete deletes a cancel func from the collection given an id.
+func (d *Deadlines) Delete(id string) {
+	d.update <- deadline{id: id, fn: nil}
+}
+
+type deadline struct {
+	id string
+	dl time.Time
+	fn context.CancelFunc
+}
+
+// insertDeadline inserts a deadline into a given sorted list of deadlines.
+func insertDeadline(dls []deadline, dl deadline) (res []deadline) {
+	if len(dls) == 0 {
+		res = append(dls, dl)
+	} else {
+		var i int
+		for i = 0; i < len(dls); i++ {
+			if dls[i].dl.After(dl.dl) {
+				break
+			}
+		}
+		res = append(dls, dl)
+		copy(res[i+1:], res[i:])
+		res[i] = dl
+	}
+	return
+}
```

Comment:
or use `sort.Search`:

```suggestion
func insertDeadline(dls []deadline, dl deadline) (res []deadline) {

	i := sort.Search(len(dls), func(i int) bool { return dls[i].dl.After(dl.dl) })

	res = append(dls, dl)
	copy(res[i+1:], res[i:])
	res[i] = dl

	return
}
```